WIP: new externalterm API

bettio opened the pull request:
**WIP, THIS PR IS STILL MEANT ONLY FOR DISCUSSION**

These changes are made under both the "Apache 2.0" and the "GNU Lesser General
Public License 2.1 or later" license terms (dual license).

SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later

bettio commented:
1. Remove externalterm_from_const_literal. This API is used only from module_load_literal and it doesn't allow to handle out of memory in a graceful way. `src/libAtomVM/opcodesswitch.h` can already handle this kind of error, but we are aborting in module.c

2. We still miss support for a `safe` option in `binary_to_term`

3. Add a new API that can be more flexible

```C
enum ExternalTermResult
{
OutOfMem,
[...]
};

enum ExternalTermOpts
{
    Copy,
    Safe
    ...
};

ExternalTermResult result externalterm_heap_from_buf(const void *buf, size_t buf_size, ExternalTermOpts opts, Heap *out_heap, term *out_term, GlobalContext *glb);

ExternalTermResult result externalterm_term_from_buf(const void *buf, size_t buf_size, ExternalTermOpts opts, term roots[], term roots_count, term *out_term, Context *ctx);
```

pguyot commented on src/libAtomVM/externalterm.h:
```diff
@@ -45,21 +45,127 @@ enum ExternalTermResult
+    GlobalContext *glb)
+{
+    const uint8_t *external_term_buf = (const uint8_t *) buf;
+    if (UNLIKELY(external_term_buf[0] != EXTERNAL_TERM_TAG)) {
```
We don't check that buf_size > 0 here.

bettio commented:
I had time to do some automated changes to the external term API, in addition to the previous changes: I made it compliant with the `C_CODING_STYLE.md` and added missing doc.
Each commit has an extensive description of which rule violation we had, and how I fixed it.

Since there is a `git mv` involved, diff is unreliable and it looks all green, so I suggest reviewing commits.

petermm commented:
the code:

# PR Review — externalterm API rework (commits 1–5)

**Commits reviewed:**
1. `be88a5add` — externalterm: add new deserialization APIs
2. `0d3282784` — Replace externalterm\_from\_binary with externalterm\_deserialize\_buf APIs
3. `cf827860b` — nifs.c: remove redundant check
4. `d99458896` — Fix externalterm\_serialize\_term parameters ordering
5. `65f31cbda` — Rename ExternalTermResult enum to external\_term\_write\_result\_t + doc

**Files changed:** `externalterm.h`, `externalterm.c`, `nifs.c`, `dist_nifs.c`

---

**3 issues found**

| # | Severity | Location | Problem | Why | Suggestion |
|---|----------|----------|---------|-----|------------|
| 1 | MEDIUM | `externalterm.h` `externalterm_compute_external_size()` doc | Doc comment says `@param size the required buffer size (tag excluded)` but the function returns `raw_size + 1`, i.e. tag **included**. The `_raw` variant's doc correctly says "tag excluded". | Callers allocating buffers based on this doc would under-allocate by 1 byte. `dist_enqueue_message` uses this function and allocates `1 + control_message_size + payload_size` — if a future caller trusts the doc and doesn't add 1 for the tag, they'll have a 1-byte buffer overrun. | Change doc to `@param size the required buffer size (tag included)` or `@param size the required buffer size` |
| 2 | LOW | `externalterm.c:106` `externalterm_deserialize_buf_raw` | `buf_size` is accepted but `UNUSED`. The validate/deserialize contract requires calling validate first, so this is by design — but the parameter can't serve as a safety net if a caller skips validation. | Documented via `@warning`, so correct by contract. However, the asymmetry with the validate function (which does use `buf_size`) could confuse future contributors. | Consider a comment explaining the design choice, e.g. `/* buf_size is unused: validation already bounds-checked */` instead of bare `UNUSED(buf_size)`. |
| 3 | LOW | `externalterm.c:142` `externalterm_from_binary_with_roots` | On OOM, the `fprintf` uses `eterm_size` (bytes read) instead of `heap_usage` (the value actually passed to `memory_ensure_free_with_roots`). | The error message reports the wrong number — confusing when debugging OOM issues. This is pre-existing but worth fixing while the API is being reworked. | Change to `fprintf(stderr, "Unable to ensure %zu free words in heap\n", (size_t) heap_usage);` |

---

### Positive observations

- **GC-safety improvement in `nif_erlang_binary_to_term`**: The old code called `externalterm_from_binary` which internally triggered GC, then had a second `memory_ensure_free_with_roots` for the `{Term, BytesUsed}` tuple (rooting `dst` but not `binary`). The new code does a single allocation (`required_heap + TUPLE_SIZE(2)`) with `binary` rooted, then deserializes with a valid pointer. This is both safer and more efficient.
- **Clean validate/deserialize split**: The two-step API (`validate_buf` → `deserialize_buf`) with `_raw` variants for tag-excluded buffers is well-designed. It lets callers like `dist_nifs.c` (which strips the tag externally) use the raw variants directly.
- **`buf_size < 1` guard in `externalterm_validate_buf`**: Correctly prevents the out-of-bounds read and unsigned underflow that would occur with an empty buffer. The old `externalterm_from_binary_with_roots` had an equivalent guard (`binary_len <= offset`), and this new API preserves that safety.
- **`*bytes_read` only written on success**: Avoids the uninitialized-read UB that would occur if `*bytes_read = raw_bytes_read + 1` were unconditional.
- **Parameter reordering** (`term` before `buf` in serialize functions): Follows a consistent "subject first, destination second" convention that matches the rest of the codebase.
- **Enum/type renames**: `ExternalTermResult` → `external_term_write_result_t` with `typedef enum` follows the `_t` suffix convention used elsewhere in AtomVM.
- **`argc` check removal**: Safe — NIF dispatch guarantees arity via gperf tables.
- **Thorough Doxygen documentation** with `@warning` preconditions on the deserialize functions.
